### PR TITLE
Update torch TP device map.

### PR DESCRIPTION
**Pull request by grimoire**
Update device map when TP.

- If host memory is enough, The model would be loaded on the host and distributed to devices.
- If these is no enough host memory, a device map would be created to load the weight on available devices and distribute them.

### CODE REVIEW

**Comment by lvhan028** [comment]
What's the risk of using `auto` device map?

**Comment by grimoire** [comment]
1. All available devices would be used, and it is hard to clear all caches in unexpected devices.
2. nearby weights would be dispatched to the same device. For example, if we have `[w0, w1, w2, w3]`, `auto` would tend to put `[w0, w1]` in device0 and `[w2, w3]` in device1. Which might leads to OOM when distributing weights.

File: lmdeploy/pytorch/engine/model_agent.py [hunk 1]
```diff
@@ -566,6 +566,32 @@ def raise_error(self, default_error: Exception):
...
+                       world_size: int,
+                       device_map: dict = None):
+    """Distribute params to each devices."""
+    if device_map is None:
+        device_map = dict()
+    device_id = 0
+    for name, _ in model.named_parameters():
+        device_map[name] = device_id
+        device_id = (device_id + 1) % world_size
+    for name, _ in model.named_buffers():
```
Comment (lvhan028) [inline lmdeploy/pytorch/engine/model_agent.py:589]: Will this method make some GPUs heavily loaded and others are not?

Comment (grimoire) [reply lmdeploy/pytorch/engine/model_agent.py:589]: In some special case, yes.
For example, if `tp==2` and weight size is `[large, small, large, small, ...]`, device0 would load all large weights and device1 would load all small weights. But I think this is almost not possible for LLM.

File: lmdeploy/pytorch/engine/model_agent.py [hunk 2]
```diff
@@ -628,18 +665,24 @@ def _broadcast_config(cache_config):
...
+            if rank == 0:
+                # adapter would remove weight of linear.
+                device_map = _create_device_map(model, world_size, device_map)
         model.eval()
         model.config.use_cache = True
 
         if rank == 0:
             with LoadNoInit():
-                device_map = 'auto'
+                device_map = __get_device_map(model, device_map)
```
Comment (pppppM) [inline lmdeploy/pytorch/engine/model_agent.py:685]: Assuming there are two gpus and two linears, is the ﻿`device_map` on ﻿rank0 `{'linear0':0, 'linear1':1}`?

If the GPU memory of ﻿`linear0` + `﻿linear1` is greater than the available GPU memory of rank0, but `﻿linear0` and ﻿`linear1` are individually less than the available GPU memory of rank0 and rank1 respectively, will `﻿device_map` be switched to the CPU?

Comment (grimoire) [reply lmdeploy/pytorch/engine/model_agent.py:685]: > Assuming there are two gpus and two linears, is the ﻿device_map on ﻿rank0 {'linear0':0, 'linear1':1}?

Yes.

> will ﻿device_map be switched to the CPU?

t0:
- rank0: `linear0`
- rank1: `linear1`

t1
- rank0: `linear0/2`
- rank1: `linear0/2` + `linear1`

t2
- rank0: `linear0/2` + `linear1/2`
- rank1: `linear0/2` + `linear1/2`

Max memory usage would be  `linear0/2` + `linear1` on rank1

No CPU fallback.

**Review by pppppM — APPROVED** [review]

**Pull request merged**